File: python/tests/_fixtures_profile.py
"""Test-support profile for the shared conformance fixtures (design §3.2).

Registers the kinds the fixture corpora use; not shipped API. `zzz` stays
unregistered on purpose — the check oracle pins `unknown-kind` for it.
"""
from __future__ import annotations

from pydantic import BaseModel, ConfigDict, Field
from pydantic import ValidationError as PydanticValidationError

from nodes.kernel.errors import FacetError, InvariantError
from nodes.kernel.node import Node
from nodes.kernel.registry import KindSpec, Registry

SOURCE = "source"
NOTE = "note"
TOPIC = "topic"
PAPER = "paper"


class Source(BaseModel):
    """Bibliographic facet the fixture `paper` nodes carry."""

    model_config = ConfigDict(extra="forbid")  # unknown keys (typos) fail, never silently dropped

    authors: list[str] = Field(default_factory=list)
    year: int | None = None
    container: str | None = None
    identifier: str | None = None
    url: str | None = None


def source_of(node: Node) -> Source:
    raw = node.facets.get(SOURCE)
    if raw is None:
        raise FacetError(f"{node.id}: missing '{SOURCE}' facet")
    try:
        return Source.model_validate(raw)
    except PydanticValidationError as exc:  # malformed payload / unknown key / wrong type
        raise FacetError(f"{node.id}: invalid '{SOURCE}' facet: {exc}") from exc


def require_identifiable_source(node: Node) -> None:
    s = source_of(node)
    if not (s.authors or s.year or s.identifier or s.url):
        raise InvariantError(
            f"{node.id}: source facet needs at least one of authors/year/identifier/url"
        )


def register_fixtures_profile(reg: Registry) -> None:
    reg.register(KindSpec(name=NOTE))
    reg.register(KindSpec(name=TOPIC))
    reg.register(
        KindSpec(
            name=PAPER,
            required_facets={SOURCE},
            invariants=[require_identifiable_source],
        )
    )

File: ts/tests/fixtures-profile.ts
import { z } from "zod";
import { FacetError, InvariantError } from "../src/errors.js";
import type { Node } from "../src/node.js";
import type { Registry } from "../src/registry.js";

/** Test-support profile for the shared conformance fixtures (design §3.2).
 *  Registers the kinds the fixture corpora use; not shipped API. */

export const SOURCE = "source";

const SourceYearSchema = z.preprocess((value) => {
  if (typeof value === "string" && value.trim() !== "") return Number(value);
  return value;
}, z.number().int().nullable().default(null));

/** `.strict()` mirrors Pydantic's `extra="forbid"`: unknown keys (typos) fail,
 *  never silently dropped. */
export const SourceSchema = z
  .object({
    authors: z.array(z.string()).default([]),
    year: SourceYearSchema,
    container: z.string().nullable().default(null),
    identifier: z.string().nullable().default(null),
    url: z.string().nullable().default(null),
  })
  .strict();

export type Source = z.infer<typeof SourceSchema>;

export function sourceOf(node: Node): Source {
  const raw = node.facets[SOURCE];
  if (raw === undefined) {
    throw new FacetError(`${node.id}: missing '${SOURCE}' facet`);
  }
  try {
    return SourceSchema.parse(raw);
  } catch (e) {
    if (e instanceof z.ZodError) {
      throw new FacetError(`${node.id}: invalid '${SOURCE}' facet: ${e.issues.map((i) => i.message).join("; ")}`);
    }
    throw e;
  }
}

export function requireIdentifiableSource(node: Node): void {
  const s = sourceOf(node);
  // Truthiness on purpose: an empty author list / year 0 / "" identifier counts as absent.
  if (!(s.authors.length || s.year || s.identifier || s.url)) {
    throw new InvariantError(`${node.id}: source facet needs at least one of authors/year/identifier/url`);
  }
}

export const NOTE = "note";
export const TOPIC = "topic";
export const PAPER = "paper";

/** Register the kinds the shared fixtures need. `zzz` stays unregistered on
 *  purpose — the check oracle pins `unknown-kind` for it. */
export function registerFixturesProfile(reg: Registry): void {
  reg.register({ name: NOTE });
  reg.register({ name: TOPIC });
  reg.register({
    name: PAPER,
    requiredFacets: new Set([SOURCE]),
    invariants: [requireIdentifiableSource],
  });
}
